File: ios/SiezaGym/Models/Exercise.swift
import Foundation

/// Grupos musculares del catalogo. Los rawValue son las claves exactas que
/// guarda Firestore en `muscleWeights`, no se pueden renombrar sin migrar data.
nonisolated enum MuscleGroup: String, CaseIterable, Sendable, Hashable {
    case pecho
    case dorsal
    case espaldaAltaTrapecio
    case deltoideAnterior
    case deltoideLateral
    case deltoidePosterior
    case biceps
    case triceps
    case antebrazo
    case cuadriceps
    case isquiotibiales
    case gluteo
    case aductores
    case gemelo
    case abdomen
    case lumbar

    var label: String {
        switch self {
        case .pecho: "Pecho"
        case .dorsal: "Dorsal"
        case .espaldaAltaTrapecio: "Espalda alta y trapecio"
        case .deltoideAnterior: "Deltoide anterior"
        case .deltoideLateral: "Deltoide lateral"
        case .deltoidePosterior: "Deltoide posterior"
        case .biceps: "Bíceps"
        case .triceps: "Tríceps"
        case .antebrazo: "Antebrazo"
        case .cuadriceps: "Cuádriceps"
        case .isquiotibiales: "Isquiotibiales"
        case .gluteo: "Glúteo"
        case .aductores: "Aductores"
        case .gemelo: "Gemelo"
        case .abdomen: "Abdomen"
        case .lumbar: "Lumbar"
        }
    }
}

nonisolated enum Equipment: String, Sendable {
    case barra, mancuerna, maquina, polea, pesoCorporal = "peso_corporal", banda, kettlebell

    var label: String {
        switch self {
        case .barra: "Barra"
        case .mancuerna: "Mancuerna"
        case .maquina: "Máquina"
        case .polea: "Polea"
        case .pesoCorporal: "Peso corporal"
        case .banda: "Banda"
        case .kettlebell: "Kettlebell"
        }
    }
}

nonisolated enum MovementPattern: String, Sendable {
    case empujeHorizontal = "empuje_horizontal"
    case empujeVertical = "empuje_vertical"
    case traccionHorizontal = "traccion_horizontal"
    case traccionVertical = "traccion_vertical"
    case dominanteRodilla = "dominante_rodilla"
    case dominanteCadera = "dominante_cadera"
    case aislamiento
    case core

    var isPush: Bool { self == .empujeHorizontal || self == .empujeVertical }
    var isPull: Bool { self == .traccionHorizontal || self == .traccionVertical }

    var label: String {
        switch self {
        case .empujeHorizontal: "Empuje horizontal"
        case .empujeVertical: "Empuje vertical"
        case .traccionHorizontal: "Tracción horizontal"
        case .traccionVertical: "Tracción vertical"
        case .dominanteRodilla: "Dominante de rodilla"
        case .dominanteCadera: "Dominante de cadera"
        case .aislamiento: "Aislamiento"
        case .core: "Core"
        }
    }
}

nonisolated enum RegistrationType: String, Sendable {
    case pesoReps = "peso_reps"
    case reps
    case tiempo
    case distanciaTiempo = "distancia_tiempo"

    /// En estos, `targetReps` es una duracion en segundos, no una cantidad.
    var isTimeBased: Bool { self == .tiempo || self == .distanciaTiempo }

    var label: String {
        switch self {
        case .pesoReps: "Peso × reps"
        case .reps: "Solo reps"
        case .tiempo: "Tiempo"
        case .distanciaTiempo: "Distancia + tiempo"
        }
    }
}

nonisolated struct Exercise: Identifiable, Sendable, Hashable {
    let id: String
    let nameEs: String
    let nameEn: String
    let equipment: Equipment?
    let pattern: MovementPattern?
    /// Reparto del esfuerzo entre musculos. Los valores suman 1.0.
    let muscleWeights: [MuscleGroup: Double]
    let registrationType: RegistrationType
    let unilateral: Bool
    let descriptionEs: String
    let mediaURL: URL?

    var primaryMuscle: MuscleGroup? {
        muscleWeights.max { $0.value < $1.value }?.key
    }
}

nonisolated extension Exercise {
    init(id: String, data: [String: Any]) {
        self.id = id
        nameEs = data["nameEs"] as? String ?? id
        nameEn = data["nameEn"] as? String ?? ""
        equipment = (data["equipment"] as? String).flatMap(Equipment.init(rawValue:))
        pattern = (data["pattern"] as? String).flatMap(MovementPattern.init(rawValue:))
        registrationType = (data["registrationType"] as? String)
            .flatMap(RegistrationType.init(rawValue:)) ?? .pesoReps
        unilateral = FirestoreValue.bool(data["unilateral"]) ?? false
        descriptionEs = data["descriptionEs"] as? String ?? ""
        mediaURL = (data["mediaUrl"] as? String).flatMap(URL.init(string:))

        var weights: [MuscleGroup: Double] = [:]
        for (key, value) in data["muscleWeights"] as? [String: Any] ?? [:] {
            guard let muscle = MuscleGroup(rawValue: key), let share = FirestoreValue.double(value) else { continue }
            weights[muscle] = share
        }
        muscleWeights = weights
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/Models/Exercise.kt
package com.siezagym.app.Models

import com.siezagym.app.Services.FirestoreValue

/** Grupos musculares del catálogo. Los rawValue son las claves exactas que
 *  guarda Firestore en `muscleWeights`, no se pueden renombrar sin migrar data. */
enum class MuscleGroup(val raw: String, val label: String) {
    PECHO("pecho", "Pecho"),
    DORSAL("dorsal", "Dorsal"),
    ESPALDA_ALTA_TRAPECIO("espaldaAltaTrapecio", "Espalda alta y trapecio"),
    DELTOIDE_ANTERIOR("deltoideAnterior", "Deltoide anterior"),
    DELTOIDE_LATERAL("deltoideLateral", "Deltoide lateral"),
    DELTOIDE_POSTERIOR("deltoidePosterior", "Deltoide posterior"),
    BICEPS("biceps", "Bíceps"),
    TRICEPS("triceps", "Tríceps"),
    ANTEBRAZO("antebrazo", "Antebrazo"),
    CUADRICEPS("cuadriceps", "Cuádriceps"),
    ISQUIOTIBIALES("isquiotibiales", "Isquiotibiales"),
    GLUTEO("gluteo", "Glúteo"),
    ADUCTORES("aductores", "Aductores"),
    GEMELO("gemelo", "Gemelo"),
    ABDOMEN("abdomen", "Abdomen"),
    LUMBAR("lumbar", "Lumbar");

    companion object {
        fun fromRaw(value: String?): MuscleGroup? = entries.firstOrNull { it.raw == value }
    }
}

enum class Equipment(val raw: String, val label: String) {
    BARRA("barra", "Barra"),
    MANCUERNA("mancuerna", "Mancuerna"),
    MAQUINA("maquina", "Máquina"),
    POLEA("polea", "Polea"),
    PESO_CORPORAL("peso_corporal", "Peso corporal"),
    BANDA("banda", "Banda"),
    KETTLEBELL("kettlebell", "Kettlebell");

    companion object {
        fun fromRaw(value: String?): Equipment? = entries.firstOrNull { it.raw == value }
    }
}

enum class MovementPattern(val raw: String, val label: String) {
    EMPUJE_HORIZONTAL("empuje_horizontal", "Empuje horizontal"),
    EMPUJE_VERTICAL("empuje_vertical", "Empuje vertical"),
    TRACCION_HORIZONTAL("traccion_horizontal", "Tracción horizontal"),
    TRACCION_VERTICAL("traccion_vertical", "Tracción vertical"),
    DOMINANTE_RODILLA("dominante_rodilla", "Dominante de rodilla"),
    DOMINANTE_CADERA("dominante_cadera", "Dominante de cadera"),
    AISLAMIENTO("aislamiento", "Aislamiento"),
    CORE("core", "Core");

    val isPush: Boolean get() = this == EMPUJE_HORIZONTAL || this == EMPUJE_VERTICAL
    val isPull: Boolean get() = this == TRACCION_HORIZONTAL || this == TRACCION_VERTICAL

    companion object {
        fun fromRaw(value: String?): MovementPattern? = entries.firstOrNull { it.raw == value }
    }
}

enum class RegistrationType(val raw: String, val label: String) {
    PESO_REPS("peso_reps", "Peso × reps"),
    REPS("reps", "Solo reps"),
    TIEMPO("tiempo", "Tiempo"),
    DISTANCIA_TIEMPO("distancia_tiempo", "Distancia + tiempo");

    /** En estos, `targetReps` es una duración en segundos, no una cantidad. */
    val isTimeBased: Boolean get() = this == TIEMPO || this == DISTANCIA_TIEMPO

    companion object {
        fun fromRaw(value: String?): RegistrationType? = entries.firstOrNull { it.raw == value }
    }
}

enum class ExerciseSource(val raw: String) {
    CATALOG("catalog"),
    CUSTOM("custom");
}

data class Exercise(
    val id: String,
    val nameEs: String,
    val nameEn: String,
    val equipment: Equipment?,
    val pattern: MovementPattern?,
    /** Reparto del esfuerzo entre músculos. Los valores suman 1.0. */
    val muscleWeights: Map<MuscleGroup, Double>,
    val registrationType: RegistrationType,
    val unilateral: Boolean,
    val descriptionEs: String,
    val mediaUrl: String?,
    val source: ExerciseSource,
) {
    val primaryMuscle: MuscleGroup?
        get() = muscleWeights.maxByOrNull { it.value }?.key

    companion object {
        fun fromRawValue(id: String, data: Map<String, Any?>, source: ExerciseSource = ExerciseSource.CATALOG): Exercise {
            var weights: LinkedHashMap<MuscleGroup, Double> = LinkedHashMap()
            (data["muscleWeights"] as? Map<*, *>)?.forEach { (key, value) ->
                val muscle = MuscleGroup.fromRaw(key as? String)
                val share = FirestoreValue.double(value)
                if (muscle != null && share != null) weights[muscle] = share
            }
            return Exercise(
                id = id,
                nameEs = data["nameEs"] as? String ?: id,
                nameEn = data["nameEn"] as? String ?: "",
                equipment = Equipment.fromRaw(data["equipment"] as? String),
                pattern = MovementPattern.fromRaw(data["pattern"] as? String),
                muscleWeights = weights,
                registrationType = RegistrationType.fromRaw(data["registrationType"] as? String)
                    ?: RegistrationType.PESO_REPS,
                unilateral = FirestoreValue.bool(data["unilateral"]) ?: false,
                descriptionEs = data["descriptionEs"] as? String ?: "",
                mediaUrl = data["mediaUrl"] as? String,
                source = source,
            )
        }
    }
}